File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.cc
#include <bits/stdc++.h>

using namespace std;

int Map[26][26];
int visited[26][26];
int dx[] = { 0,0,-1,1 };
int dy[] = { 1,-1,0,0 };
int n;

typedef struct node {
	int x;
	int y;
};

queue<node> q;
vector<int> v;

int bfs(int i, int j)
{
	int val = 0;

	q.push({ i,j });
	visited[i][j] = 1;

	while (!q.empty())
	{
		int cx = q.front().x;
		int cy = q.front().y;

		val++;

		q.pop();

		for (int i = 0; i < 4; i++)
		{
			int nx = cx + dx[i];
			int ny = cy + dy[i];

			if (nx < 0 || ny < 0 || nx >= n || ny >= n || visited[nx][ny] || !Map[nx][ny] )
				continue;

			else
			{
				q.push({ nx,ny });
				visited[nx][ny] = 1;
			}
		}
	}

	return val;
}

int main()
{
	cin >> n;

	for (int i = 0; i < n; i++)
	{
		string str;
		cin >> str;
		for (int j = 0; j < str.size(); j++)
			Map[i][j] = str[j] - '0';
	}

	for (int i = 0; i < n; i++)
	{
		for (int j = 0; j < n; j++)
		{
			if (Map[i][j] && !visited[i][j])
			{
				int result = bfs(i, j);
				v.push_back(result);
			}
		}
	}

	sort(v.begin(), v.end());

	cout << v.size() << "\n";

	for (int i = 0; i < v.size(); i++)
	{
		cout << v[i] << "\n";
	}

	return 0;
}

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.util.*;
import java.io.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();

    static int n;
    static int[][] map;
    static boolean[][] visited;
    static int[] dx = {0,0,-1,1};
    static int[] dy = {1,-1,0,0};
    static ArrayList<Integer> list = new ArrayList<>();


    public static void main(String[] args) throws Exception {

        n = Integer.parseInt(br.readLine());

        map = new int[n][n];
        visited = new boolean[n][n];

        for (int i = 0 ; i < n ; i ++) {
            String str = br.readLine();
            for (int j = 0 ; j < n ; j ++) {
                map[i][j] = str.charAt(j) - '0';
            }
        }

        for (int i = 0 ; i < n ; i ++) {
            for (int j = 0 ; j < n ; j ++) {
                if (!visited[i][j] && map[i][j] == 1) {
                    list.add(bfs(i,j));
                }
            }
        }

        Collections.sort(list);

        System.out.println(list.size());

        for (int result: list) {
            System.out.println(result);
        }
    }

    static class Node {
        int x;
        int y;

        Node(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    static int bfs(int x, int y) {

        visited[x][y] = true;

        ArrayDeque<Node> q = new ArrayDeque<Node>();

        q.add(new Node(x, y));

        int count = 1;

        while(!q.isEmpty()) {
            Node cur = q.poll();

            int cx = cur.x;
            int cy = cur.y;

            for (int i = 0 ; i < 4 ; i ++) {
                int nx = cx + dx[i];
                int ny = cy + dy[i];

                if (nx < 0 || ny < 0 || nx >= n || ny >= n) continue;

                if (!visited[nx][ny] && map[nx][ny] == 1) {
                    q.add(new Node(nx, ny));
                    visited[nx][ny] = true;
                    count ++;
                }
            }
        }

        return count;
    }
}